Allow SyncBatchNorm without DDP in inference mode

ppwwyyxx opened the pull request:
Summary: Fix https://github.com/pytorch/pytorch/issues/22538

Differential Revision: D16883694

mrshenli commented on torch/nn/modules/batchnorm.py:
```diff
@@ -440,19 +437,24 @@ def forward(self, input):
+            process_group = torch.distributed.group.WORLD
+            if self.process_group:
+                process_group = self.process_group
+            world_size = torch.distributed.get_world_size(process_group)
```
[`process_group.size()`](https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/c10d/init.cpp#L220) should work as well.

ppwwyyxx replied:
From the implementation of `get_group_size` it looks like `.size()` is not sufficient?

https://github.com/pytorch/pytorch/blob/dfdb86a59577d8b0fc4565988a8ac01b5ecd339f/torch/distributed/distributed_c10d.py#L194-L204

mrshenli commented on torch/nn/modules/batchnorm.py:
```diff
@@ -440,19 +437,24 @@ def forward(self, input):
+            if self.process_group:
+                process_group = self.process_group
+            world_size = torch.distributed.get_world_size(process_group)
+            need_sync = need_sync and world_size > 1
```
`need_sync = world_size > 1` should be sufficient, as otherwise, it wouldn't enter this clause.

mrshenli reviewed (commented):
Thanks for fixing this! Shall we add a test for it (SyncBatchNorm without DDP)?

mrshenli reviewed (approved):
Hey @ppwwyyxx, thanks for adding the fix. Feel free to land if you need this feature urgently, but would really prefer to have a test for it.

ppwwyyxx committed:
Allow SyncBatchNorm without DDP in inference mode (#24815)

Summary:
Fix https://github.com/pytorch/pytorch/issues/22538
Pull Request resolved: https://github.com/pytorch/pytorch/pull/24815

Test Plan:
Can run a detectron2 evaluation without entering DDP.

#sandcastle

Differential Revision: D16883694

fbshipit-source-id: a9199a82a55ba784a9d80005969cccf32d6b4827
